rts/motoko-rts/src/gc/incremental/partitioned_heap.rs

```diff
@@ -390,40 +491,125 @@ impl PartitionedHeap {
         }
     }
 
+    fn partitions(&mut self) -> PartitionIterator {
+        PartitionIterator::new(self)
+    }
+
     pub fn is_initialized(&self) -> bool {
-        self.partitions[0].index == 0
+        self.number_of_partitions > 0
     }
 
     pub fn base_address(&self) -> usize {
         self.heap_base
     }
 
+    unsafe fn get_extension_table(&self, partition_index: usize) -> *mut PartitionTable {
+        debug_assert!(partition_index >= PARTITIONS_PER_TABLE);
+        let mut index = partition_index - PARTITIONS_PER_TABLE;
+        let mut table = self.partition_table.extension;
+        while index >= PARTITIONS_PER_TABLE {
+            index -= PARTITIONS_PER_TABLE;
+            debug_assert_ne!((*table).extension, null_mut());
+            table = (*table).extension;
+        }
+        table
+    }
+
+    unsafe fn get_partition_table(&self, partition_index: usize) -> *const PartitionTable {
+        let first_table = &self.partition_table as *const PartitionTable;
+        debug_assert_ne!(first_table, null());
+        if partition_index < PARTITIONS_PER_TABLE {
+            first_table
+        } else {
+            self.get_extension_table(partition_index)
+        }
+    }
+
+    unsafe fn mutable_partition_table(&mut self, partition_index: usize) -> *mut PartitionTable {
+        let first_table = &mut self.partition_table as *mut PartitionTable;
+        debug_assert_ne!(first_table, null_mut());
+        if partition_index < PARTITIONS_PER_TABLE {
+            first_table
+        } else {
+            self.get_extension_table(partition_index)
+        }
+    }
+
     pub fn get_partition(&self, index: usize) -> &Partition {
-        &self.partitions[index]
+        unsafe {
+            let table = self.get_partition_table(index);
+            let offset = index % PARTITIONS_PER_TABLE;
+            &(*table).partitions[offset]
+        }
     }
 
     fn mutable_partition(&mut self, index: usize) -> &mut Partition {
-        &mut self.partitions[index]
+        unsafe {
+            let table = self.mutable_partition_table(index);
+            let offset = index % PARTITIONS_PER_TABLE;
+            &mut (*table).partitions[offset]
+        }
+    }
```

Comment:
NIT: These two functions are almost identical, can one express one in terms of the other?